src/rgw/rgw_dedup_cluster.cc

```diff
@@ -0,0 +1,1068 @@
+    std::map<std::string, bufferlist> set_pairs = {{RGW_DEDUP_ATTR_EPOCH, new_epoch_bl}};
+    librados::ObjectWriteOperation op;
+    ret = cmp_vals_set_vals(op, Mode::String, Op::EQ, cmp_pairs, set_pairs);
+    if (ret != 0) {
+      ldpp_dout(dpp, 5) << __func__ << "::failed cmp_vals_set_vals" << dendl;
+      return -EINVAL;
+    }
+    ldpp_dout(dpp, 20) << __func__ << "::send EPOCH CLS" << dendl;
+    ret = ioctx.operate(oid, &op);
+    if (ret == 0) {
+      ldpp_dout(dpp, 10) << __func__ << "::Epoch object was written" << dendl;
+    }
+    else if (ret == -EEXIST) {
+      ldpp_dout(dpp, 10) << __func__ << "::Accept existing Epoch object" << dendl;
+      ret = 0;
+    }
+    else {
+      ldpp_dout(dpp, 1) << __func__ << "::failed ioctx.operate() with: "
+			<< cpp_strerror(ret) << ", ret=" << ret << dendl;
+    }
+    return ret;
+  }
+
+  //---------------------------------------------------------------------------
+  static int swap_epoch(const DoutPrefixProvider *dpp,
+			librados::IoCtx &ioctx,
+			const dedup_epoch_t *p_old_epoch,
+			int dedup_type,
+			work_shard_t num_work_shards,
+			md5_shard_t num_md5_shards)
+  {
+    dedup_epoch_t new_epoch = { p_old_epoch->serial + 1, dedup_type, ceph_clock_now(),
+				num_work_shards, num_md5_shards};
+    bufferlist old_epoch_bl, new_epoch_bl;
+    encode(*p_old_epoch, old_epoch_bl);
+    encode(new_epoch, new_epoch_bl);
+    ComparisonMap cmp_pairs = {{RGW_DEDUP_ATTR_EPOCH, old_epoch_bl}};
+    std::map<std::string, bufferlist> set_pairs = {{RGW_DEDUP_ATTR_EPOCH, new_epoch_bl}};
+    librados::ObjectWriteOperation op;
+    int ret = cmp_vals_set_vals(op, Mode::String, Op::EQ, cmp_pairs, set_pairs);
+    ldpp_dout(dpp, 1) << __func__ << "::send EPOCH CLS" << dendl;
+    std::string oid(DEDUP_EPOCH_TOKEN);
+    ret = ioctx.operate(oid, &op);
+    return ret;
+  }
+
+  //---------------------------------------------------------------------------
+  struct shard_progress_t {
+    shard_progress_t() {
+      this->creation_time   = utime_t();
+      this->completion_time = utime_t();
+      this->update_time     = utime_t();
+    }
+
+    shard_progress_t(uint64_t _progress_a,
+		     uint64_t _progress_b,
+		     bool _completed,
+		     const std::string &_owner,
+		     const bufferlist  &_in_bl) :owner(_owner), in_bl(_in_bl) {
+      this->progress_a  = _progress_a;
+      this->progress_b  = _progress_b;
+      this->completed   = _completed;
+      this->update_time = ceph_clock_now();
+
+      if (_progress_a == SP_NO_OBJECTS && _progress_b == SP_NO_OBJECTS) {
+	this->creation_time = ceph_clock_now();
+      }
+      if (_completed) {
+	this->completion_time = ceph_clock_now();
+      }
+    }
+
+    uint64_t    progress_a = SP_NO_OBJECTS;
+    uint64_t    progress_b = SP_NO_OBJECTS;
+    bool        completed = false;
+    utime_t     update_time;
+    utime_t     creation_time;
+    utime_t     completion_time;
+    std::string owner;
```

Comment:
what value do you use as "owner"?

Reply:
It is left empty and will be set once we have the data.
We only create empty shard_progress_t before reading into them